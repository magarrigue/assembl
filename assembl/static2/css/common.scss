html {
  font-size: 12px;  // the base for the rem unit (used in react-datepicker)
}

body {
  background-color: $page-bkg-grey;
  font-family: $font-family;
  font-size: $font-size-s;
  letter-spacing: $text-letter-spacing;
  overflow-y: scroll;
  overflow-x: hidden;
  -ms-overflow-style: scrollbar; // prevent transparent scrollbar in IE
}

a {
  color: $link-color;
  text-decoration: none;

  &:hover {
    color: $link-hover-color;
    text-decoration: none;
  }
}

input {
  &[type='text'] {
    box-shadow: 0 0 0 50px $input-txt-bkg-color inset;

    &:focus {
      box-shadow: 0 0 0 50px $input-txt-bkg-color inset;
    }
  }

  &[type='password'] {
    box-shadow: 0 0 0 50px $input-txt-bkg-color inset;

    &:focus {
      box-shadow: 0 0 0 50px $input-txt-bkg-color inset;
    }
  }
}

input:-moz-placeholder, textarea:-moz-placeholder {
  font-family: $font-family6;
}

input::-webkit-input-placeholder, textarea::-webkit-input-placeholder {
  font-family: $font-family6;
}

input.placeholder, textarea.placeholder {
  font-family: $font-family6;
}

.overflow-x {
  overflow-x: scroll;
}

.pre-wrap {
  white-space: pre-wrap;
}

.white {
  color: $white;
}

.grey {
  color: $grey2;
}

.black {
  color: $black;
}

.color {
  color: $first-color;
}

.color2 {
  color: $second-color;
}

.app-content {
  margin: $navbar-height 0 0;
}

.max-container {
  margin: 0 auto;
  max-width: $app-max-width;
}

.background-light {
  background-color: $page-bkg-light;
}

.background-grey {
  background-color: $page-bkg-grey;
}

.background-dark-grey {
  background-color: $page-bkg-dark-grey;
}

.background-dark {
  background-color: $page-bkg-dark;
}

.background-color {
  background-color: $first-color-light;
}

.background-img {
  background-position: top left;
  background-repeat: no-repeat;
  background-size: cover;
}

.content-section {
  padding: $margin-section 0;
}

.mobile-content-section {
  padding-top: 40px !important;
}

.content-margin {
  margin: -30px 0 0;
}

.inline {
  display: inline-block;
}

.left {
  float: left;
}

.right {
  float: right;
}

.center {
  text-align: center;
  margin-left: auto;
  margin-right: auto;
}

.justify {
  text-align: justify;
}

.border-top {
  border-top: $navbar-border;
}

.border-right {
  border-right: $navbar-border;
}

.border-left {
  border-left: $navbar-border;
}

.border-bottom {
  border-bottom: $navbar-border;
}

.border {
  border: $navbar-border;
}

.no-padding {
  padding: 0;
}

.no-margin {
  margin: 0;
}

.padding {
  padding: 15px;
}

.margin-xs {
  margin-top: 5px;
}

.margin-s {
  margin-top: 10px;
}

.margin-m {
  margin-top: 20px;
}

.margin-l {
  margin-top: 40px;
}

.margin-xl {
  margin-top: 60px;
}

.margin-xxl {
  margin-top: 80px;
}

.margin-xxxl {
  margin-top: 120px;
}

.margin-left-xxl {
  margin-left: 75px;
}

.margin-left-2 {
  margin-left: 2px;
}

.full-size {
  width: 100%;
}

.underline {
  border-bottom: $box-border;
  margin: 0 0 20px;
}

.uppercase {
  text-transform: uppercase;
}

.hidden {
  display: none;
}

.shown {
  display: block;
}

.col-centered {
  float: none;
  margin: 0 auto;
}

.clear {
  clear: both;
}

.relative {
  position: relative;
}

.pointer {
  cursor: pointer;
}

.no-pointer-events {
  pointer-events: none;
}

.form-label {
  color: $grey3;
  font-family: $font-family;
  font-size: $font-size-xs;
  padding: 0 0 0 5px;
}

.form-label-body {
  position: absolute;
  top: 131px;
}

.form-tree > .form-container > .form-tree-item {
  &::after,
  &::before {
    display: none;
  }
}

.form-tree {
  .form-container:first-child > .form-tree-item::before {
    display: none;
  }

  .form-container:first-child > .form-tree-item::after {
    border-radius: 5px 0 0 0;
  }

  .plus.form-tree-item:last-child::after,
  .form-container:last-child > .form-tree-item::after {
    display: none;
  }

  .plus.form-tree-item:last-child::before,
  .form-container:last-child > .form-tree-item::before {
    border-radius: 0 0 0 5px;
  }

  .form-container {
    .form-group {
      margin-bottom: 10px !important;
    }

    .form-tree-item {
      position: relative;

      &.plus {
        margin: initial;
        height: 15px;
        width: 15px;
        font-size: 12px;
      }
    }

    .form-tree-item::after {
      position: absolute;
      content: "";
      top: 18px;
      left: -15px;
      border-left: 1px solid #ccc;
      border-top: 1px solid #ccc;
      border-radius: 0 0 0 0;
      width: 15px;
      height: 100%;
    }

    .form-tree-item.plus::after {
      top: 8px;
    }

    .form-tree-item::before {
      content: "";
      position: absolute;
      top: -26px;
      left: -15px;
      border-left: 1px solid #ccc;
      border-bottom: 1px solid #ccc;
      border-radius: 0 0 0 0;
      width: 15px;
      height: 45px;
    }

    .form-tree-item.plus::before {
      height: 19px;
      top: -10px;
    }
  }

  .form-branch {
    margin-left: 25px;
    padding-right: 25px;
  }
}

.capitalized {
  text-transform: capitalize;
}

@media screen and (min-width: 992px) {
  .col-bottom {
    display: table-cell;
    float: none;
    vertical-align: bottom;
  }
}

@media screen and (max-width: 420px) {
  .content-section {
    padding: $margin-section-xs 0;
  }
}

strong {
  font-family: $font-family3;
}

#pDebug {
  display: none;
}

.cursor-wait {
  cursor: wait !important;
}

@media screen and (min-width: 1024px) {
  .page-body {
    margin: 0 90px 80px 90px;
  }
}

.small-hyphen-padding {
  padding: 0 5px;
}

.jumbotron {
  margin-top: 50px;
}

.flex {
  display: flex;
}

.center-flex {
  display: flex;
  justify-content: center;
}
