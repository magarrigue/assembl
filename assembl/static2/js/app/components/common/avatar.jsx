import React from 'react';
import { connect } from 'react-redux';
import { compose, graphql } from 'react-apollo';
import { Translate } from 'react-redux-i18n';
import { Link } from 'react-router';
import { NavDropdown, MenuItem } from 'react-bootstrap';

import { getContextual, get } from '../../utils/routeMap';
import UserQuery from '../../graphql/userQuery.graphql';
import withoutLoadingIndicator from './withoutLoadingIndicator';
import { browserHistory } from '../../router';

class ProfileIcon extends React.Component {
  constructor(props) {
    super(props);
    this.setCurrentView = this.setCurrentView.bind(this);
  }

  componentWillMount() {
    this.setCurrentView();
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.location) {
      this.setState({ next: nextProps.location });
    }
  }

  setCurrentView() {
    const { location } = this.props;
    this.setState({
      next: location
    });
  }

  render() {
    const { slug, connectedUserId } = this.props;
    const dropdownUser = (
      <div className="inline">
        <span className="assembl-icon-profil grey" />
        <span className="user-account"><Translate value="profile.panelTitle" /></span>
      </div>
    );
    return (
      <div className="right avatar">
        {!connectedUserId && (
          <Link to={`${getContextual('login', { slug: slug })}?next=${this.state.next}`}>
            <div className="connection">
              <Translate value="navbar.connection" />
            </div>
          </Link>
        )}
        {connectedUserId && (
          <ul className="dropdown-xs">
            <NavDropdown pullRight title={dropdownUser} id="user-dropdown">
              <MenuItem
                onClick={() => {
                  browserHistory.push(get('profile', { slug: slug, userId: connectedUserId }));
                }}
              >
                <Translate value="profile.panelTitle" />
              </MenuItem>
              <MenuItem href={`${getContextual('oldLogout', { slug: slug })}?next=${get('home', { slug: slug })}`}>
                <Translate value="navbar.logout" />
              </MenuItem>
            </NavDropdown>
          </ul>
        )}
      </div>
    );
  }
}

const mapStateToProps = ({ context, debate }) => ({
  slug: debate.debateData.slug,
  connectedUserId: context.connectedUserId,
  id: context.connectedUserIdBase64
});

export default compose(
  connect(mapStateToProps),
  graphql(UserQuery, {
    props: ({ data }) => {
      if (data.loading) {
        return { loading: true };
      }
      if (data.error) {
        return { error: data.error };
      }
      return {
        displayName: data.user.displayName
      };
    }
  }),
  withoutLoadingIndicator()
)(ProfileIcon);