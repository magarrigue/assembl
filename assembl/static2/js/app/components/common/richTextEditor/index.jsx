// @flow
import RichTextEditor from './new';
// import RichTextEditor from './old';

export default RichTextEditor;