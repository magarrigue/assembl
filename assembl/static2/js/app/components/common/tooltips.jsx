// @flow

import * as React from 'react';
import { Tooltip } from 'react-bootstrap';
import { Translate } from 'react-redux-i18n';

type TooletipProps = {
  level: number
};

export const addSectionTooltip = (
  <Tooltip id="addSectionTooltip">
    <Translate value="administration.sections.addSection" />
  </Tooltip>
);

export const deleteSectionTooltip = (
  <Tooltip id="deleteSectionTooltip">
    <Translate value="administration.sections.deleteSection" />
  </Tooltip>
);

export const addVoteProposalTooltip = (
  <Tooltip id="addProposalTooltip">
    <Translate value="administration.voteProposals.addProposal" />
  </Tooltip>
);

export const deleteVoteProposalTooltip = (
  <Tooltip id="deleteProposalTooltip">
    <Translate value="administration.voteProposals.deleteProposal" />
  </Tooltip>
);

export const upTooltip = (
  <Tooltip id="upTooltip">
    <Translate value="administration.up" />
  </Tooltip>
);

export const downTooltip = (
  <Tooltip id="downTooltip">
    <Translate value="administration.down" />
  </Tooltip>
);

export const addThematicTooltip = ({ level }: TooletipProps) => (
  <Tooltip id="addThematicTooltip">
    <Translate value="administration.addThematic" level={level} />
  </Tooltip>
);

export const addQuestionTooltip = () => (
  <Tooltip id="addQuestionTooltip">
    <Translate value="administration.addQuestion" />
  </Tooltip>
);

export const deleteQuestionTooltip = () => (
  <Tooltip id="deleteQuestionTooltip">
    <Translate value="administration.deleteQuestion" />
  </Tooltip>
);

export const deleteThematicTooltip = () => (
  <Tooltip id="deleteThematicTooltip">
    <Translate value="administration.deleteThematic" />
  </Tooltip>
);

export const languageTooltip = (
  <Tooltip id="languageTooltip">
    <Translate value="administration.changeLanguage" />
  </Tooltip>
);

export const answerTooltip = (
  <Tooltip id="answerTooltip">
    <Translate value="debate.toAnswer" />
  </Tooltip>
);

export const sharePostTooltip = (
  <Tooltip id="sharePostTooltip">
    <Translate value="debate.sharePost" />
  </Tooltip>
);

export const shareSynthesisTooltip = (
  <Tooltip id="shareSynthesisTooltip">
    <Translate value="debate.shareSynthesis" />
  </Tooltip>
);

export const likeTooltip = (
  <Tooltip id="likeTooltip">
    <Translate value="debate.like" />
  </Tooltip>
);

export const disagreeTooltip = (
  <Tooltip id="disagreeTooltip">
    <Translate value="debate.disagree" />
  </Tooltip>
);

export const dontUnderstandTooltip = (
  <Tooltip id="dontUnderstandTooltip">
    <Translate value="debate.dontUnderstand" />
  </Tooltip>
);

export const moreInfoTooltip = (
  <Tooltip id="moreInfoTooltip">
    <Translate value="debate.moreInfo" />
  </Tooltip>
);

export const deleteMessageTooltip = (
  <Tooltip id="deleteMessageTooltip">
    <Translate value="debate.deleteMessage" />
  </Tooltip>
);

export const editMessageTooltip = (
  <Tooltip id="editMessageTooltip">
    <Translate value="debate.editMessage" />
  </Tooltip>
);

export const createResourceTooltip = () => (
  <Tooltip id="createResourceTooltip">
    <Translate value="administration.resourcesCenter.createResource" />
  </Tooltip>
);

export const deleteResourceTooltip = () => (
  <Tooltip id="deleteResourceTooltip">
    <Translate value="administration.resourcesCenter.deleteResource" />
  </Tooltip>
);

export const resetTokensTooltip = (
  <Tooltip id="resetTokensTooltip">
    <Translate value="debate.voteSession.resetTokens" />
  </Tooltip>
);

export const hiddenTooltip = <Tooltip id="hiddenTooltip" style={{ display: 'none' }} />;

export const notEnoughTokensTooltip = (
  <Tooltip id="notEnoughTokensTooltip">
    <Translate value="debate.voteSession.notEnoughTokens" />
  </Tooltip>
);

export const exclusiveTokensTooltip = (
  <Tooltip id="exclusiveTokensTooltip">
    <Translate value="debate.voteSession.exclusiveTokens" />
  </Tooltip>
);

export const nextStepTooltip = (
  <Tooltip id="nextStepTooltip">
    <Translate value="administration.nextStep" />
  </Tooltip>
);

export const previousStepTooltip = (
  <Tooltip id="previousStepTooltip">
    <Translate value="administration.previousStep" />
  </Tooltip>
);

export const validateExtractTooltip = (
  <Tooltip id="validateExtractTooltip">
    <Translate value="harvesting.validateExtract" />
  </Tooltip>
);

export const editExtractTooltip = (
  <Tooltip id="editExtractTooltip">
    <Translate value="harvesting.editExtract" />
  </Tooltip>
);

export const deleteExtractTooltip = (
  <Tooltip id="deleteExtractTooltip">
    <Translate value="harvesting.deleteExtract" />
  </Tooltip>
);

export const nuggetExtractTooltip = (
  <Tooltip id="nuggetExtractTooltip">
    <Translate value="harvesting.nuggetExtract" />
  </Tooltip>
);

export const qualifyExtractTooltip = (
  <Tooltip id="qualifyExtractTooltip">
    <Translate value="harvesting.qualifyExtract" />
  </Tooltip>
);

export const harvestingTooltip = (
  <Tooltip id="harvestingTooltip">
    <Translate value="harvesting.harvesting" />
  </Tooltip>
);

export const deleteFileTooltip = (
  <Tooltip id="deleteFileTooltip">
    <Translate value="administration.deleteAssociatedFile" />
  </Tooltip>
);

export const deleteThematicImageTooltip = (
  <Tooltip id="deleteThematicImageTooltip">
    <Translate value="administration.deleteThematicImage" />
  </Tooltip>
);

export const addTextFieldTooltip = (
  <Tooltip id="addTextFieldTooltip">
    <Translate value="administration.profileOptions.addTextField" />
  </Tooltip>
);

export const hideTextFieldTooltip = (
  <Tooltip id="hideTextFieldTooltip">
    <Translate value="administration.profileOptions.hideTextField" />
  </Tooltip>
);

export const addSelectFieldOptionTooltip = (
  <Tooltip id="addSelectFieldTooltip">
    <Translate value="administration.profileOptions.addSelectFieldOption" />
  </Tooltip>
);

export const deleteTextFieldTooltip = (
  <Tooltip id="deleteTextFieldTooltip">
    <Translate value="administration.profileOptions.deleteTextField" />
  </Tooltip>
);

export const deleteSelectFieldOptionTooltip = (
  <Tooltip id="deleteSelectFieldOptionTooltip">
    <Translate value="administration.profileOptions.deleteSelectFieldOption" />
  </Tooltip>
);

export const textFieldToggleOptionalTooltip = (
  <Tooltip id="textFieldToggleOptionalTooltip">
    <Translate value="administration.profileOptions.textFieldToggleOptional" />
  </Tooltip>
);
export const textFieldToggleRequiredTooltip = (
  <Tooltip id="textFieldToggleRequiredTooltip">
    <Translate value="administration.profileOptions.textFieldToggleRequired" />
  </Tooltip>
);

export const addPhaseTooltip = (
  <Tooltip id="addPhaseTooltip">
    <Translate value="administration.timelineAdmin.addPhase" />
  </Tooltip>
);

export const deletePhaseTooltip = (
  <Tooltip id="deletePhaseTooltip">
    <Translate value="administration.timelineAdmin.deletePhase" />
  </Tooltip>
);

export const fileNameTooltip = (fileName?: string) => <Tooltip id="fileNameTooltip">{fileName}</Tooltip>;