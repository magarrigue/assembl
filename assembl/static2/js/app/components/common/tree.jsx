/* eslint react/no-multi-comp: "off" */

import React from 'react';
import debounce from 'lodash/debounce';
import { AutoSizer, CellMeasurer, CellMeasurerCache, List, WindowScroller } from 'react-virtualized';
import { scrollToPost } from '../../utils/hashLinkScroll';
import NuggetsManager from './nuggetsManager';

class Child extends React.PureComponent {
  static defaultProps = {
    InnerComponentFolded: () => null
  };

  constructor(props) {
    super(props);
    this.renderToggleLink = this.renderToggleLink.bind(this);
    this.expandCollapse = this.expandCollapse.bind(this);
    this.resizeTreeHeight = this.resizeTreeHeight.bind(this);
    this.scrollToElement = this.scrollToElement.bind(this);
    this.state = {
      expanded: true,
      visible: false
    };
  }

  componentDidMount() {
    this.onScroll();
    window.addEventListener('scroll', this.onScroll);
    window.addEventListener('resize', this.onScroll);
  }

  componentWillUnmount() {
    this.stopListening();
  }

  onScroll = debounce(() => {
    const holder = this.holder;
    if (!holder) {
      return;
    }
    const box = holder.getBoundingClientRect();
    const pageYOffset = window.pageYOffset;
    const top = box.top + pageYOffset;
    // visible if the top of the box is in viewport or next page
    const isVisible = top < pageYOffset + 2 * window.innerHeight && top > pageYOffset;
    if (isVisible) {
      this.setState(() => ({
        visible: true
      }));
      this.stopListening();
    }
  }, 100);

  stopListening() {
    window.removeEventListener('scroll', this.onScroll);
    window.removeEventListener('resize', this.onScroll);
  }

  resizeTreeHeight(delay = 0) {
    // This function will be called by each post rendered, so we delay the
    // recomputation until no post are rendered in 200ms to avoid unnecessary lag.
    const { listRef, cache, rowIndex, nuggetsManager } = this.props;
    cache.clear(rowIndex, 0);
    if (listRef) {
      let delayedRecomputeRowHeights = listRef.delayedRecomputeRowHeights;
      if (!delayedRecomputeRowHeights) {
        delayedRecomputeRowHeights = [null, null]; // [timeoutId, minRowIndex from which to recompute row heights]
        listRef.delayedRecomputeRowHeights = delayedRecomputeRowHeights;
      }
      if (delayedRecomputeRowHeights[0]) {
        clearTimeout(delayedRecomputeRowHeights[0]);
      }
      delayedRecomputeRowHeights[1] = Math.min(delayedRecomputeRowHeights[1] || rowIndex, rowIndex);
      delayedRecomputeRowHeights[0] = setTimeout(() => {
        // if listRef.Grid is null, it means it has been unmounted, so we are now on a new List
        if (listRef.Grid) {
          // In Firefox (tested on version 59), recomputing row heights can jump back the page scroll
          // to the same post (The scrollTop from the Grid component is fine.),
          // potentially a post with a youtube video, but may be a coincidence.
          // Saving pageYOffset and restoring it after recomputeRowHeights fixes the issue.
          const pageYOffset = window.pageYOffset;
          listRef.recomputeRowHeights(delayedRecomputeRowHeights[1]);
          window.scrollTo({ top: pageYOffset, left: 0 });
          nuggetsManager.update();
          // recompute height only for rows (top post) starting at rowIndex
        }
        delayedRecomputeRowHeights[0] = null;
        delayedRecomputeRowHeights[1] = null;
      }, delay);
    }
  }

  expandCollapse(event) {
    event.stopPropagation();
    this.setState(
      state => ({ expanded: !state.expanded }),
      () => {
        this.resizeTreeHeight(0);
      }
    );
  }

  renderToggleLink(expanded, indented) {
    return (
      <div
        ref={(el) => {
          this.scrollAnchor = el;
        }}
        onClick={(event) => {
          if (expanded) {
            this.scrollToElement();
          }
          this.expandCollapse(event);
        }}
        className={indented ? 'expand-indented' : 'expand'}
      >
        {expanded ? <span className="assembl-icon-minus-circled" /> : <span className="assembl-icon-plus-circled" />}
      </div>
    );
  }

  scrollToElement() {
    scrollToPost(this.scrollAnchor);
  }

  render() {
    const {
      contentLocaleMapping,
      hidden,
      id,
      lang,
      children,
      InnerComponent,
      InnerComponentFolded,
      level,
      originalLocale,
      rowIndex, // the index of the row (i.e. level 0 item) in the List
      SeparatorComponent,
      fullLevel,
      nuggetsManager,
      listRef,
      cache,
      identifier
    } = this.props;
    const cssClasses = () => {
      let cls = `level level-${level}`;
      if (level > 0) {
        cls += ' border-left child-level';
      }
      if (level > 3) {
        cls += ' no-shift';
      }
      if (level > 4) {
        cls += ' padding-right';
      }
      if (hidden) {
        cls += ' hidden';
      }
      return cls;
    };
    const numChildren = children ? children.length : 0;
    const expanded = this.state.expanded;
    const contentLocale = contentLocaleMapping.getIn([id, 'contentLocale'], originalLocale);
    const forwardProps = {
      contentLocale: contentLocale,
      ...this.props,
      numChildren: numChildren
    };
    delete forwardProps.children;
    // InnerComponent, the post, is only rendered when the Child appears in the viewport or next page
    const { hash } = window.location;
    let visible = this.state.visible;
    // load right away the shared post
    let hashid;
    if (hash !== '') {
      hashid = hash.replace('#', '').split('?')[0];
      if (hashid === id) {
        visible = true;
      }
    }

    return (
      <div
        className={cssClasses()}
        id={id}
        ref={(el) => {
          this.holder = el;
          if (el && hashid === id) {
            scrollToPost(el, false);
          }
        }}
      >
        {visible ? (
          <InnerComponent {...forwardProps} measureTreeHeight={this.resizeTreeHeight} />
        ) : (
          <div style={{ height: 0.5 * window.innerHeight }} />
        )}
        {numChildren > 0 ? this.renderToggleLink(expanded, level < 4) : null}
        {numChildren > 0
          ? children.map((child, idx) => {
            const fullLevelArray = 'fullLevel' in this.props ? fullLevel.split('-') : [];
            fullLevelArray[level] = idx;
            return (
              <Child
                hidden={!expanded}
                key={child.id}
                {...child}
                contentLocaleMapping={contentLocaleMapping}
                lang={lang}
                rowIndex={rowIndex}
                level={level + 1}
                InnerComponent={InnerComponent}
                InnerComponentFolded={InnerComponentFolded}
                SeparatorComponent={SeparatorComponent}
                fullLevel={fullLevelArray.join('-')}
                nuggetsManager={nuggetsManager}
                listRef={listRef}
                cache={cache}
                identifier={identifier}
              />
            );
          })
          : null}
        {numChildren > 0 && !expanded ? (
          <div
            className="postfolded-container"
            onClick={(event) => {
              if (expanded) {
                this.scrollToElement();
              }
              this.expandCollapse(event);
            }}
          >
            <div className="post-folded">
              <InnerComponentFolded nbPosts={numChildren} />
            </div>
          </div>
        ) : null}
      </div>
    );
  }
}
Child.defaultProps = {
  level: 0,
  hidden: false
};

class Tree extends React.Component {
  constructor(props) {
    super(props);
    this.nuggetsManager = new NuggetsManager();
    this.prevStopIndex = 0;
    this.cache = new CellMeasurerCache({
      defaultHeight: 500,
      minHeight: 185,
      fixedWidth: true
    });
  }

  componentDidMount() {
    // Reset the global prevStopIndex to not overfetch posts when changing idea
    // or to avoid recreating all dom nodes if we go back to the same idea.
    this.cache.clearAll();
    this.prevStopIndex = 0;

    if (this.props.initialRowIndex !== null) {
      this.listRef.scrollToRow(this.props.initialRowIndex);
    }
  }

  componentWillReceiveProps(nextProps) {
    if (this.props.data.length !== nextProps.data.length) {
      this.cache.clearAll();
      // If a new top post has been created, clear the cache
      // and rerender the List by changing its key
      // to be sure to recalculate the heights of all top posts.
      // Reset prevStopIndex because globalList.recomputeRowHeights() only update the 10 first top posts
      // because Grid.recomputeGridSize that is called by globalList.recomputeRowHeights
      // doesnt take into account our hack in overscanIndicesGetter.
      // This means that the posts after the 10th post are removed from the dom
      // and will be recreated when scrolling, losing the previous expand/collapse local state.
      this.prevStopIndex = 0;
    }
  }

  // override overscanIndicesGetter to not remove from the dom the posts once rendered
  // to fix various issue with scrolling with WindowScroller
  overscanIndicesGetter = ({ cellCount, overscanCellsCount, stopIndex }) => {
    let overscanStopIndex;
    if (cellCount === 1) {
      // overscanIndicesGetter is called for columns, not rows
      // use default implementation
      overscanStopIndex = Math.min(cellCount - 1, stopIndex + overscanCellsCount);
    } else {
      this.prevStopIndex = Math.max(this.prevStopIndex, stopIndex);
      overscanStopIndex = Math.min(cellCount - 1, this.prevStopIndex + overscanCellsCount);
    }
    return {
      overscanStartIndex: 0,
      overscanStopIndex: overscanStopIndex
    };
  };

  cellRenderer = ({ index, key, parent, style }) => {
    const {
      data,
      InnerComponent, // component that will be rendered in the child
      InnerComponentFolded, // component that will be used to render the children when folded
      SeparatorComponent, // separator component between first level children
      identifier
    } = this.props;

    const childData = data[index];
    return (
      <CellMeasurer cache={this.cache} columnIndex={0} key={key} parent={parent} rowIndex={index}>
        <div style={style}>
          {index > 0 ? <SeparatorComponent /> : null}
          <Child
            {...childData}
            contentLocaleMapping={parent.props.contentLocaleMapping}
            key={childData.id}
            lang={parent.props.lang}
            rowIndex={index}
            InnerComponent={InnerComponent}
            InnerComponentFolded={InnerComponentFolded}
            SeparatorComponent={SeparatorComponent}
            nuggetsManager={this.nuggetsManager}
            listRef={this.listRef}
            cache={this.cache}
            identifier={identifier}
          />
        </div>
      </CellMeasurer>
    );
  };

  render() {
    const { contentLocaleMapping, data, lang, noRowsRenderer } = this.props;
    return (
      <WindowScroller>
        {({ height, isScrolling, onChildScroll, scrollTop }) => (
          <AutoSizer
            disableHeight
            onResize={() => {
              this.cache.clearAll();
              this.listRef.recomputeRowHeights();
              this.nuggetsManager.update();
            }}
          >
            {({ width }) => (
              <List
                contentLocaleMapping={contentLocaleMapping}
                height={height}
                // pass lang to the List component to ensure that the rows are rendered again if we change the site language
                lang={lang}
                isScrolling={isScrolling}
                onScroll={onChildScroll}
                scrollTop={scrollTop}
                autoHeight
                rowHeight={this.cache.rowHeight}
                deferredMeasurementCache={this.cache}
                noRowsRenderer={noRowsRenderer}
                ref={(ref) => {
                  // Add a guard because the List component's ref is recalculated many times onScroll
                  // Causing other components that have listRef as prop to re-render 4x-6x times.
                  if (!this.listRef) {
                    this.listRef = ref;
                  }
                }}
                rowCount={data.length}
                overscanIndicesGetter={this.overscanIndicesGetter}
                overscanRowCount={1}
                rowRenderer={this.cellRenderer}
                width={width}
                className="tree-list"
              />
            )}
          </AutoSizer>
        )}
      </WindowScroller>
    );
  }
}

export default Tree;