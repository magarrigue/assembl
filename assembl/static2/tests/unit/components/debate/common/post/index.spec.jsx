import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import { DumbPost } from '../../../../../../js/app/components/debate/common/post';

const handleEditClickSpy = jest.fn();
const refetchIdeaSpy = jest.fn();
export const postProps = {
  body: '<p>You can\'t index the port without programming the wireless HTTP program!</p>',
  borderLeftColor: 'red',
  contentLocale: 'fr',
  creationDate: '11/11/2011',
  data: {
    post: {
      creator: {
        userId: 'dannietreutel',
        displayName: 'Dannie Treutel'
      },
      id: 'XYZ333',
      dbId: 124,
      modified: true,
      sentimentCounts: 3,
      mySentiment: 'like',
      attachments: [],
      extracts: [],
      bodyMimeType: 'text/html',
      bodyEntries: [
        {
          localeCode: 'en',
          value: 'I\'ll bypass the online CSS bus, that should firewall the SAS monitor!'
        }
      ],
      publicationState: 'PUBLISHED',
      subjectEntries: [
        {
          localeCode: 'en',
          value: 'We need to hack the solid state HTTP alarm!'
        }
      ]
    }
  },
  debateData: {
    translationEnabled: true
  },
  editable: true,
  fullLevel: '0',
  id: 'XYZ333',
  ideaId: 'ABC123',
  identifier: 'thread',
  phaseId: 'DEF345',
  indirectIdeaContentLinks: [{ idea: { id: 'foo', title: 'Foo' } }],
  lang: 'fr',
  level: 1,
  modifiedSubject: <span>open-source Associate</span>,
  originalLocale: 'fr',
  publicationState: 'PUBLISHED',
  refetchIdea: refetchIdeaSpy,
  rowIndex: 0,
  subject: 'open-source Associate',
  timeline: [],
  handleEditClick: handleEditClickSpy,
  nuggetsManager: null,
  routerParams: {
    phase: 'thread',
    slug: '/debate/thread/XYZ333',
    themeId: 'VVV999'
  },
  numChildren: 3
};

describe('Post component', () => {
  it('should render a deleted post', () => {
    const props = {
      data: {
        post: {
          bodyEntries: [
            {
              localeCode: 'en',
              value: 'I\'ll bypass the online CSS bus, that should firewall the SAS monitor!'
            }
          ],
          id: 'XYZ123',
          publicationState: 'DELETED_BY_USER',
          subjectEntries: [
            {
              localeCode: 'en',
              value: 'We need to hack the solid state HTTP alarm!'
            }
          ]
        }
      }
    };
    const renderer = new ShallowRenderer();
    renderer.render(<DumbPost {...props} />);
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  // TODO: find a way to set state or to call handleEditClick
  it('should render a form to edit the post if mode is edit');

  it('should render a PostView component and pass its props to it', () => {
    const props = postProps;
    const renderer = new ShallowRenderer();
    renderer.render(<DumbPost {...props} />);
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
});