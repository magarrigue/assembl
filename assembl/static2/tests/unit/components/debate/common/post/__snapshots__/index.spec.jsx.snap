// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post component should render a PostView component and pass its props to it 1`] = `
<div
  className="posts"
  id="XYZ333"
>
  <PostView
    body="I'll bypass the online CSS bus, that should firewall the SAS monitor!"
    borderLeftColor="red"
    contentLocale="fr"
    creationDate="11/11/2011"
    data={
      Object {
        "post": Object {
          "attachments": Array [],
          "bodyEntries": Array [
            Object {
              "localeCode": "en",
              "value": "I'll bypass the online CSS bus, that should firewall the SAS monitor!",
            },
          ],
          "bodyMimeType": "text/html",
          "creator": Object {
            "displayName": "Dannie Treutel",
            "userId": "dannietreutel",
          },
          "dbId": 124,
          "extracts": Array [],
          "id": "XYZ333",
          "modified": true,
          "mySentiment": "like",
          "publicationState": "PUBLISHED",
          "sentimentCounts": 3,
          "subjectEntries": Array [
            Object {
              "localeCode": "en",
              "value": "We need to hack the solid state HTTP alarm!",
            },
          ],
        },
      }
    }
    debateData={
      Object {
        "translationEnabled": true,
      }
    }
    editable={true}
    fullLevel="0"
    handleEditClick={[Function]}
    id="XYZ333"
    ideaId="ABC123"
    identifier="thread"
    indirectIdeaContentLinks={
      Array [
        Object {
          "idea": Object {
            "id": "foo",
            "title": "Foo",
          },
        },
      ]
    }
    isHarvesting={false}
    lang="fr"
    level={1}
    modifiedSubject={
      <span>
        <span
          className="subject-prefix"
        >
          Rep. 1: 
        </span>
        We need to hack the solid state HTTP alarm!
      </span>
    }
    multiColumns={false}
    nuggetsManager={null}
    numChildren={3}
    originalLocale="fr"
    phaseId="DEF345"
    publicationState="PUBLISHED"
    refetchIdea={[Function]}
    routerParams={
      Object {
        "phase": "thread",
        "slug": "/debate/thread/XYZ333",
        "themeId": "VVV999",
      }
    }
    rowIndex={0}
    subject="We need to hack the solid state HTTP alarm!"
    timeline={Array []}
  />
</div>
`;

exports[`Post component should render a deleted post 1`] = `
<DeletedPost
  deletedBy="user"
  id={undefined}
  subject={
    <span>
      We need to hack the solid state HTTP alarm!
    </span>
  }
/>
`;
